fix(surveys): Preserve survey choices in a question when switching to compatible question type

lshaowei18 opened the pull request:
## Problem
Closes #36530

## Changes

1) When switching between two question types with choices (`MultipleChoice`& `SingleChoice`), the choices will be retained
2) When switching between question types with choices to question types without choices, a confirmation dialog will be shown. If cancel is clicked, we explicitly set the type back to the original type. This is because by the time we receive `onSelect`, the type has already been switched. (There is another event handler called `onChange`, but that is inserted by `LemonField` and i'm not sure how to update it)


### Switching to type with choices


https://github.com/user-attachments/assets/1475cfdd-b938-48e6-8672-d386ff19f34b

### Switching to other types

https://github.com/user-attachments/assets/79793f29-9c5b-4cc2-9afd-b1ec4ece1d4d







## How did you test this code?

Locally!

1) Go to http://localhost:8010/project/1/surveys/new?edit=true
2) Update the first question's type to single choice and add some additional choices
3) Switch the type to multiple choice (You should see that the choices are not removed)
4) Switch the type to rating
5) You should see a confirmation prompt 
6) Click on cancel (your question should remain unchanged)
7) Try switching again, this time round click on ok in the confirmation prompt
8) You should see that the question is now updated to the new type

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

lshaowei18 committed:
fix: allow survey to switch question type without resetting

lshaowei18 committed:
fix: ensure that types with choices will ask for confirmation when switching to types with no choices

lshaowei18 commented on frontend/src/scenes/surveys/SurveyEditQuestionRow.tsx:
```diff
@@ -164,6 +168,27 @@ export function SurveyEditQuestionGroup({ index, question }: { index: number; qu
+                            if (isCurrentTypeMultipleSurveyQuestion && !isNewTypeMultipleSurveyQuestion) {
+                                if (
+                                    !confirm(
+                                        'The choices you have configured will be removed. Would you like to proceed?'
```
Hey @benHPostHog, do you mind helping me out with this confirmation message? Would love to hear any suggestions of yours!

<img width="680" height="282" alt="Screenshot 2025-08-15 at 6 34 31 AM" src="https://github.com/user-attachments/assets/2a1654ae-b0d7-4b47-a769-b70ab742b9ce" />

lshaowei18 replied:
I updated the UI to use lemon dialog, let me know if you have any suggestions on the title and message :) 

<img width="1118" height="328" alt="Screenshot 2025-08-15 at 8 06 41 AM" src="https://github.com/user-attachments/assets/fe51043b-9c2a-4c8d-b6cf-0123c38d84cc" />

lshaowei18 commented on frontend/src/scenes/surveys/SurveyEditQuestionRow.tsx:
```diff
@@ -164,6 +168,27 @@ export function SurveyEditQuestionGroup({ index, question }: { index: number; qu
+                            }
+                            if (isCurrentTypeMultipleSurveyQuestion && !isNewTypeMultipleSurveyQuestion) {
+                                if (
+                                    !confirm(
+                                        'The choices you have configured will be removed. Would you like to proceed?'
+                                    )
+                                ) {
+                                    // Doing this because by the time we receive `onSelect`, the question type has already changed
+                                    setQuestionType(index, question.type)
+                                    return
+                                }
```
Good point!

lshaowei18 committed:
fix: question to come from component instead of relying on existing state

lshaowei18 committed:
fix: switch to a modal

lshaowei18 committed:
fix: add question to setDefaultForQuestionType

lucasheriques reviewed (commented):
overall logic is ok, but that `onSelect` function is getting kind of big. maybe we can extract that into helper methods that could be easier to read? what do you think?

here's what I'm thinking:

```tsx
// outside the component
const isMultipleChoiceQuestion = (type: SurveyQuestionType): boolean =>
    type === SurveyQuestionType.MultipleChoice || type === SurveyQuestionType.SingleChoice

const getEditingFlags = (
    question: SurveyQuestion,
    survey: Survey | NewSurvey
): {
    editingQuestion: boolean
    editingDescription: boolean
    editingThankYouMessage: boolean
} => ({
    editingQuestion: defaultSurveyFieldValues[question.type].questions[0].question !== question.question,
    editingDescription: defaultSurveyFieldValues[question.type].questions[0].description !== question.description,
    editingThankYouMessage:
        defaultSurveyFieldValues[question.type].appearance.thankYouMessageHeader !==
        survey.appearance?.thankYouMessageHeader,
})
```

```tsx
// inside the component
const confirmQuestionTypeChange = (
    index: number,
    question: SurveyQuestion,
    newType: SurveyQuestionType,
    editingFlags: ReturnType<typeof getEditingFlags>
): void => {
    // Reset to current type first (because onSelect has already changed it)
    setMultipleSurveyQuestion(index, question, question.type)

    LemonDialog.open({
        title: 'Changing question type',
        description: (
            <p className="py-2">The choices you have configured will be removed. Would you like to proceed?</p>
        ),
        primaryButton: {
            children: 'Continue',
            status: 'danger',
            onClick: () => {
                setDefaultForQuestionType(
                    index,
                    question,
                    newType,
                    editingFlags.editingQuestion,
                    editingFlags.editingDescription,
                    editingFlags.editingThankYouMessage
                )
                resetBranchingForQuestion(index)
            },
        },
        secondaryButton: {
            children: 'Cancel',
        },
    })
}


onSelect={(newType) => {
    const isCurrentMultipleChoice = isMultipleChoiceQuestion(question.type)
    const isNewMultipleChoice = isMultipleChoiceQuestion(newType)

    const editingFlags = getEditingFlags(question, survey)

    // Same multiple choice type - just update type
    if (isCurrentMultipleChoice && isNewMultipleChoice) {
        setMultipleSurveyQuestion(index, question, newType)
        resetBranchingForQuestion(index)
        return
    }

    // Changing from multiple choice - warn about losing choices
    if (isCurrentMultipleChoice && !isNewMultipleChoice) {
        confirmQuestionTypeChange(index, question, newType, editingFlags)
        return
    }

    // Default case - simple type change
    setDefaultForQuestionType(
        index,
        question,
        newType,
        editingFlags.editingQuestion,
        editingFlags.editingDescription,
        editingFlags.editingThankYouMessage
    )
    resetBranchingForQuestion(index)
}}
```

what do you think?

lshaowei18 committed:
fix: move booleans into logic, refactor questionRow onSelect

lshaowei18 commented:
Hey @lucasheriques; love the feedback!

I added one change where i moved the `editingQuestions` booleans into `setDefaultForQuestionType`. It does make `setDefaultForQuestionType` abit harder to read.

Let me know what you think!

lshaowei18 committed:
fix: revert from `isMultipleChoiceQuestion` so that the typing can be correct

lshaowei18 committed:
fix: update typing for function

lshaowei18 committed:
Merge branch 'master' into improve-survey-question-switching

lshaowei18 committed:
Merge branch 'master' into improve-survey-question-switching

lucasheriques reviewed (approved):
looks great now, thanks for the PR 🙏

lucasheriques committed:
Merge branch 'master' into improve-survey-question-switching
